src/Network/Wai/Log/Options.hs

```diff
@@ -2,26 +2,29 @@ module Network.Wai.Log.Options (
 -- * Options & Timing
   Options(..)
 , ResponseTime(..)
+, logSendingResponse
 -- * Defaults
 , defaultOptions
 , defaultLogRequest
 , defaultLogResponse
 ) where
 
-import Data.Aeson.Types (Pair)
+import Data.Aeson.Types (Pair, Value)
+import Data.ByteString.Builder (Builder)
 import Data.String.Conversions (ConvertibleStrings, StrictText, cs)
 import Data.Text (Text)
 import Data.Time.Clock (NominalDiffTime)
+import Data.UUID (UUID)
 import Log
 import Network.HTTP.Types.Status
 import Network.Wai
 
 -- | Logging options
 data Options = Options {
-    logLevel            :: LogLevel
-  , logRequest          :: Request -> [Pair]
-  , logSendingResponse  :: Bool
-  , logResponse         :: Request -> Response -> ResponseTime -> [Pair]
+    logLevel    :: LogLevel
+  , logRequest  :: UUID -> Request -> [Pair]
+  , logResponse :: UUID -> Request -> Response -> Value -> ResponseTime -> [Pair]
+  , logBody     :: Maybe (Status -> Builder -> Value)
```

Comment:
Right, I see why the extra `logBody` option is needed, I did not think that you need to do `IO` to get stuff out of the `Response` 😞 but of course you _do_ need to...

One problem I see with the current construction is that whenever `isJust logBody == True`, then we are performing extra work with the `Response`, even though in the vast majority of cases you probably don't want to log anything. In that respect, it would be better if we had something like:
```hs
logBody :: Maybe (Status -> Builder -> Maybe Value)
```

Although another restriction still present is you can only decide whether to log the body based on the `Status`.

Currently the most general way I can see for this library is to have a custom type:
```suggestion
  , logResponse :: LogResponse
```

With something like
```hs
-- | Log response function can either be a pure function (@LogResponseSimple@),
-- or an @IO@ function (@LogResponseIO@) for when you want to inspect the
-- 'Response' using IO
data LogResponse
  = LogResponseSimple (Request -> Response -> ResponseTime -> [Pair])
  | LogResponseIO     (Request -> Response -> ResponseTime -> IO [Pair])
```

Then in `Log.Internal` we wouldn't need to extract the `Response` but instead do
```hs
    case logResponse of
      LogResponseSimple logSimple ->
        logIO "Request complete" . object $ logSimple req resp times
      LogResponseIO logWithIO -> do
        pairs <- logWithIO req resp times
        logIO "Request complete" . object $ pairs
```

Reply:
We talked about having
```hs
Request -> Status -> ResponseHeaders -> Maybe (Builder -> Value)
```
for the optional body logger.

This should provide enough generality without forcing users of the library to convert the `Response` into a `Builder` themselves.

Reply:
Implemented the other idea we discussed.